lib/start-command.js

```diff
@@ -0,0 +1,132 @@
+  // up configuration stored in it's working directory.
+
+  if (stat.isFile()) {
+    // The app is a file, we'll run it from its module's root directory.
+    return resolveFile(cwd, script);
+  } else if (stat.isDirectory()) {
+    // The app is a directory (such as '.'), we'll look for what to run from it.
+    try {
+      require.resolve(path.resolve(cwd, script));
+      return resolveDir(cwd, script);
+    } catch (err) {
+      // given path is not resolvable as a module
+      return resolvePackageStart(cwd, script);
+    }
+  }
+
+  return app;
+}
+
+function ifExists(base, file) {
+  return fs.existsSync(path.resolve(base, file)) ? file : null;
+}
+
+function resolveFile(cwd, script) {
+  var absolute = path.resolve(cwd, script);
+  var dir = path.dirname(absolute);
+  var file = path.basename(absolute);
+  // walk up the path until we find the package.json nearest to the app
+  while (!fs.existsSync(path.join(dir, 'package.json'))) {
+    file = path.join(path.basename(dir), file);
+    // hit root, fallback to running as-is
+    if (dir === path.dirname(dir)) {
+      file = path.basename(script);
+      dir = path.dirname(script);
+      break;
+    }
+    dir = path.dirname(dir);
+  }
+  return {cwd: dir, path: file, error: null};
+}
+
+function resolveDir(cwd, dir) {
+  // The canonical way to run is with `npm start`, but we can't do that for a
+  // cluster worker, because the worker must start the runner (so it can
+  // require strong-agent). Instead, we search for a runnable file, and if
+  // that fails, just require the directory and let node do the right thing
+  // (which should be to use package.main if it exists, index.js otherwise).
+  var app = {
+    cwd: path.resolve(cwd, dir),
+    path: ifExists(cwd, 'server.js') || ifExists(cwd, 'app.js') || '.',
+    error: null,
+  };
+  return app;
+}
+
+function resolvePackageStart(cwd, script) {
+  var pkg = null;
+  var app = resolveFile(cwd, path.resolve(cwd, script, 'package.json'));
+  try {
+    pkg = JSON.parse(fs.readFileSync(path.resolve(app.cwd, app.path), 'utf8'));
+    if (pkg && pkg.scripts && pkg.scripts.start) {
+      return fromStart(app.cwd, pkg.scripts.start);
+    }
+  } catch (e) {
+    app.error = e;
+  }
+  app.error = app.error || Error('Could not determine how to start ' + script);
+  return app;
+}
+
+function fromStart(cwd, script) {
+  var app = {
+    cwd: cwd,
+    path: null,
+    error: null,
+  };
+  if (!script) {
+    app.error = Error('no script');
+    return app;
+  }
+  var scriptRegexp = /^(\S+)((?:\s+--\S+)*)?(?:\s+(\S+))(\s.*)?$/;
+  var parts = scriptRegexp.exec(script);
+  debug('script: %j => ', script, parts);
+  // TODO: support extracting node options like --expose_gc and --harmony and
+  // alternative executors, like coffee.
+  // app.execPath = parts[1];
+  // app.execArgs = parts[2];
+  app.path = ifExists(cwd, parts[3]);
+  if (!app.path) {
+    app.error = Error('Could not determine start script from ' + script);
```

Comment:
"determine main script" (I think the "start script" is package.scripts.start, its the main script or main file we can't determine from that)